### ISSUE TITLE
proposal: `sort_public_methods_before_private_methods`

### ISSUE THREAD

**Issue by ju851zel** [dart-lang/linter#3798]
## Description

Define public methods before private methods.

## Details

**DO** define public methods before private methods.

This makes the code uniform across multiple classes and it makes it faster to find specific methods in a class.
Because public methods are the interface to the class they are (often) more important and should come before private methods. 

This would be my first approach to have a defined order for different parts in a class.
[Flutter style guide ordering of class members](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#order-other-class-members-in-a-way-that-makes-sense)

## Kind

Style

## Good Examples 

```dart
class A {
  int a() => 0;
  int c() => 0;
  int _b() => 0;
}
```

## Bad Examples

```dart
class A {
  int a() => 0;
  int _b() => 0;
  int c() => 0;
}
```

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by bwilkerson** [comment]
I made a couple of comments on the PR, but we should really discuss the semantics (and desirability) of the lint before getting too far into the implementation issues.

Do we want to have a rule that enforces part of the sort order, or would it be better to have a rule that enforces a more complete sort order (such as the one produced by the "Sort Members" feature)? We do already have a couple of rules about the order of constructors, but we need to consider interactions between all such rules, and the more rules we have the harder it will be to manage the interactions. We could, for example, have a rule to enforce that "Sort Members" has been run over the code.

If we do want a more limited rule like this, how does this relate to different kinds of members? The "Sort Members" feature, for example produces the following order:
- public static fields
- private static fields
- public instance fields
- private instance fields
- public constructors
- private constructors
- public instance methods
- private instance methods
- public static methods
- private static methods

Public members of a given kind are always before private members of the same kind, but the kinds are kept separate. Does this rule propose that private members of any kind always be after public members of every kind?

**Comment by srawlins** [comment]
Possible duplicate of https://github.com/dart-lang/linter/issues/2367

**Comment by srawlins** [comment]
We will need some strong justification in order to incorporate this proposed rule. We are only accepting new style-based lint rules which are promoted by the Effective Dart style guide, or the Flutter style guide. I don't believe either recommends putting public methods before private ones.

I typically place private methods near their use, for example. Google's [Java Style Guide](https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents) recommends a logical order, but doesn't call out public-before-private, as it really is on a class-by-class basis. Our internal Java Style Guide (OK I have no idea why they're different) spells out the style I use:

> Some of these members feel like "footnotes" and some feel like "appendices". If a member exists only to support a single other method, put it just below that method ("footnote" style). But if it is (or feels like) a more general helper, then put it toward the end of the class ("appendix" style).

**Comment by srawlins** [comment]
I'll close this for now. If we have a top-priority style guide which recommends an ordering of members, we can re-consider this rule.
